src/compiler/utilities.ts

```diff
@@ -4666,8 +4666,12 @@ namespace ts {
 
     /* @internal */
     export function isNodeArray<T extends Node>(array: ReadonlyArray<T>): array is NodeArray<T> {
-        return array.hasOwnProperty("pos")
-            && array.hasOwnProperty("end");
+        const res = array.hasOwnProperty("pos") && array.hasOwnProperty("end");
+        if (res) {

```

Comment:
We possibly call this a lot between visitor and factory. Do the added checks have any perf impact?